feat: allow build-time overrides of attestation server url and static fingerprint override

cassandracomar opened the pull request:
pulls the attestation server url and overriden fingerprint from res/values/strings, allowing them to be overriden at AOSP/GrapheneOS build time. because the device info/fingerprint can't be known statically, this PR introduces a dynamic look up for the device info and fingerprint via a singleton that gets its context set by the main activity, avoiding null pointer issues by ensuring static accesses for Resources in AttestationProtocol.java occur as late as possible.

Commit:
pull attest. server url from res/values/strings

this allows the attestation server url to be overriden at
AOSP/GrapheneOS build time.

Commit:
dynamically look up device info and fingerprint

when the fingerprint override resource is set at build time, dynamically
look up the device info for the current device, checking that the AVB
fingerprint matches the one the Auditor app was built with.

uses a singleton to allow lazily-initialized lookups of the device name
and override fingerprint in a static context.

Commit:
use new properties for device name and os name

thestinger commented on app/src/main/java/app/attestation/auditor/AttestationActivity.java:
```diff
@@ -507,6 +506,10 @@ public boolean onPrepareOptionsMenu(final Menu menu) {
         return true;
     }
 
+    final String tutorial_url() {
```
Should make this `private` since Java defaults to package public. Don't need to make it `final` since if we want to do that it probably makes more sense to just make the classes `final` but it isn't really needed. It doesn't have the usual meaning of an immutable variable binding here but rather preventing an override by a class inheriting this one.

thestinger commented on app/src/main/java/app/attestation/auditor/RemoteVerifyJob.java:
```diff
@@ -127,6 +124,18 @@ static void cancel(final Context context) {
         scheduler.cancel(FIRST_RUN_JOB_ID);
     }
 
+    final String domain() {
```
Same as tutorial_url.

thestinger commented on app/src/main/res/values/strings.xml:
```diff
@@ -1,5 +1,6 @@
 <resources>
     <string name="app_name">Auditor</string>
+    <string name="url">attestation.app</string>
```
Probably better to call this `base_url` and it can include the `https://` prefix.

thestinger commented on app/src/main/res/values/strings.xml:
```diff
@@ -1,6 +1,7 @@
 <resources>
     <string name="app_name">Auditor</string>
     <string name="url">attestation.app</string>
+    <string name="avb_fingerprint_override">NOT OVERRIDEN</string>
```
This approach won't really work because Auditor requires 2 devices. The standard GrapheneOS and stock OS entries should also be left working for a custom build. Instead, new entries should be added to the non-stock OS table and the person doing it will need to use their build on both ends since our Auditor builds won't trust their Auditor builds (signing key fingerprint of the app is verified remotely via key attestation) and won't have the entries for their OS builds.

cassandracomar replied:
this is here so there's still a local check of the fingerprint before the verification is submitted. I'm unsure how new entries can be added to the table without source code modification at build time. do you have thoughts?

cassandracomar replied:
I took a pass at resolving this by setting up maps in the XML resources to connect device identifiers with fingerprints. this should allow someone building a custom os for several devices to inject their own map of keys / DeviceInfo without source code modification. as I needed to validate the parser anyway, I've moved the existing static maps into their own resource maps, but I'm open to thoughts on that -- it is useful for ensuring the parser doesn't break sometime in the future.

r3g-5z commented on app/src/main/java/app/attestation/auditor/RemoteVerifyJob.java:
```diff
@@ -127,6 +124,18 @@ static void cancel(final Context context) {
+        return getString(R.string.url);
+    }
+
+    final String challenge_url() {
```
Method names should also match the code/naming style of the rest of the codebase even if they're private methods

```suggestion
    final String challengeUrl() {
```

muhomorr commented:
Turning `RemoteVerifyJob.DOMAIN` constant into a BuildConfig constant would allow to simplify these changes a lot.
See Apps app for an example: https://github.com/GrapheneOS/Apps/blob/9bb136db0aacfbaef8c5ffe3c8980eb582087dba/app/build.gradle.kts#L50

Commit:
remove AttestationContext and fix naming

we're actually already passing the Context around so there's no need for
a singleton.

Commit:
swap to xml maps for fingerprints

this allows multiple devices to be configured the same even when they
have different AVB fingerprints, and therefore should be able to
validate each other.

Commit:
lower access modifiers.

they got increased when I was moving the tests around to figure out what
would make them actually function.
